fsspec/implementations/prefix.py

```diff
@@ -0,0 +1,135 @@
+import os
+from pathlib import Path
+from typing import Any, Iterable, Sequence, Union
+
+import fsspec
+from fsspec import AbstractFileSystem
+from fsspec.core import split_protocol
+from fsspec.utils import stringify_path
+
+
+class PrefixFileSystem(AbstractFileSystem):
+    def __init__(
+        self,
+        prefix: str,
+        filesystem: fsspec.AbstractFileSystem,
+        *args,
+        **storage_options,
+    ) -> None:
+        super().__init__(*args, **storage_options)
+        self.prefix = prefix
+        self.filesystem = filesystem
+
+    def _add_fs_prefix(self, path: Union[str, Path]) -> Union[str, Sequence[str]]:
+        if isinstance(path, (str, Path)):
+            path = stringify_path(path)
+            protocol, path = split_protocol(path)
+            path = os.path.join(self.prefix, path)
```

Comment:
I see, there are some special cases that we need to handle though (that `os.path.join` or `relpath` does for us), e.g.

- If `prefix==sep` we must not add the prefix, otherwise we will have a double separator
- if `prefix` is empty we must not add the sep
- if the path is already absolute, we should not modify it. i.e. `os.path.join('c/d', '/a/b')` returns `/a/b`.  ..or should we assume all paths are relative to the prefix regardless if they are absolute or not?

Reply:
> *should we assume all paths are relative to the prefix*

That would be my expectation - i.e. if I specified `/` with a prefix of `/tmp`, I would expect that to reference `/tmp`. Similarly, I would expect `/a/b` to reference `/tmp/a/b`

Reply:
Ok I implemented this behavior, the test `test_directories` now tests also for this

Reply:
I think prefix=empty and prefix=sep should raise an exception, or at least, a user has no right to assume they do anything useful :)

Reply:
At the moment if `prefix=empty` it defaults to `prefix=sep`. If `prefix=sep` the PrefixFileSystem should be a passthrough that does not change any path (if everything works).